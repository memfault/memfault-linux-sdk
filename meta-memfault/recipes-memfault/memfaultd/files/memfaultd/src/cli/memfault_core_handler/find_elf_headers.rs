//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::io::{Read, Seek};

use elf::program_header::{ProgramHeader, PT_NOTE};
use eyre::Result;
use log::debug;

use crate::cli::memfault_core_handler::core_elf_note::{iterate_elf_notes, ElfNote};
use crate::cli::memfault_core_handler::core_reader::{CoreReader, CoreReaderImpl};
use crate::cli::memfault_core_handler::memory_range::MemoryRange;
use crate::cli::memfault_core_handler::{elf, ElfPtrSize};

/// Detects whether the supplied stream is at an ELF file and if so, returns the ranges in the
/// stream, that contain the ELF header, program headers and build id note. The ranges are
/// relative to the stream's position when calling the function.
/// GDB requires the headers and note to be present, for debuginfod and build id based symbol file
/// lookups to work.
pub fn find_elf_headers_and_build_id_note_ranges<P: Read + Seek>(
    vaddr_base: ElfPtrSize,
    stream: &mut P,
) -> Result<Vec<MemoryRange>> {
    debug!(
        "Detecting ELF headers and build ID note ranges from vaddr 0x{:x}",
        vaddr_base
    );

    let mut elf_reader = CoreReaderImpl::new(stream)?;
    let elf_header = elf_reader.elf_header();
    let program_headers = elf_reader.read_program_headers()?;
    let build_id_note_ph = program_headers
        .iter()
        .find(|ph| ph.p_type == PT_NOTE && contains_gnu_build_id_note(&mut elf_reader, ph));

    match build_id_note_ph {
        Some(build_id_note_ph) => {
            let ranges = [
                // ELF header:
                MemoryRange::from_start_and_size(vaddr_base, elf_header.e_ehsize as ElfPtrSize),
                // Program header table:
                MemoryRange::from_start_and_size(
                    vaddr_base + elf_header.e_phoff,
                    (elf_header.e_phentsize as ElfPtrSize) * (elf_header.e_phnum as ElfPtrSize),
                ),
                // Note segment containing GNU Build ID:
                MemoryRange::from_start_and_size(
                    vaddr_base + build_id_note_ph.p_offset,
                    build_id_note_ph.p_filesz,
                ),
            ];

            // FIXME: MFLT-11635 CoreElf .py requires ELF header + build ID in single segment
            Ok(vec![MemoryRange::new(
                ranges
                    .iter()
                    .map(|r| r.start)
                    .min()
                    .ok_or_else(|| eyre::eyre!("No memory regions found"))?,
                ranges
                    .iter()
                    .map(|r| r.end)
                    .max()
                    .ok_or_else(|| eyre::eyre!("No memory regions found"))?,
            )])
        }
        None => Err(eyre::eyre!("Build ID note missing")),
    }
}

fn contains_gnu_build_id_note<P: Read + Seek>(
    elf_reader: &mut CoreReaderImpl<&mut P>,
    program_header: &ProgramHeader,
) -> bool {
    match elf_reader.read_segment_data(program_header) {
        Ok(data) => iterate_elf_notes(&data).any(|note| matches!(note, ElfNote::GnuBuildId(_))),
        _ => false,
    }
}

#[cfg(test)]
mod test {
    use std::fs::File;
    use std::io::{Cursor, SeekFrom};
    use std::path::PathBuf;

    use super::*;

    #[test]
    fn test_err_result_if_no_elf() {
        let mut stream = Cursor::new(vec![0u8; 100]);
        assert_eq!(
            find_elf_headers_and_build_id_note_ranges(0, &mut stream)
                .unwrap_err()
                .to_string(),
            "Invalid ELF header"
        );
    }

    #[test]
    fn test_err_result_if_missing_build_id() {
        // A core.elf itself does not have a build id note:
        let mut stream = core_elf_stream();
        assert_eq!(
            find_elf_headers_and_build_id_note_ranges(0, &mut stream)
                .unwrap_err()
                .to_string(),
            "Build ID note missing"
        );
    }

    #[test]
    #[ignore = "FIXME: MFLT-11635 CoreElf .py requires ELF header + build ID in single segment"]
    fn test_ok_result_with_list_of_ranges() {
        let mut stream = exe_elf_stream();
        assert_eq!(
            find_elf_headers_and_build_id_note_ranges(0x1000, &mut stream).unwrap(),
            // Note: see offsets/sizes from the readelf -Wl output in exe_elf_stream().
            vec![
                // ELF header
                MemoryRange::from_start_and_size(0x1000, 0x40),
                // Program header table:
                MemoryRange::from_start_and_size(0x1000 + 0x40, 0x2d8),
                // Note segment with GNU Build ID
                MemoryRange::from_start_and_size(0x1000 + 0x358, 0x44),
            ]
        );
    }

    #[test]
    fn test_ok_result_with_list_of_ranges_mflt_11635_work_around() {
        let mut stream = exe_elf_stream();
        assert_eq!(
            find_elf_headers_and_build_id_note_ranges(0x1000, &mut stream).unwrap(),
            // Note: see offsets/sizes from the readelf -Wl output in exe_elf_stream().

            // FIXME: MFLT-11635 CoreElf .py requires ELF header + build ID in single segment
            vec![MemoryRange::new(0x1000, 0x1000 + 0x358 + 0x44),]
        );
    }

    fn core_elf_stream() -> File {
        File::open(core_elf_path()).unwrap()
    }

    fn exe_elf_stream() -> File {
        let mut stream = core_elf_stream();
        // The coredump fixture contains part of a captured ELF executable at offset 0x2000:
        //   Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
        //   ...
        //   LOAD           0x002000 0x00005587ae8bd000 0x0000000000000000 0x001000 0x001000 R   0x1000

        // The embedded, partial ELF executable has this program table:
        //   Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
        //   PHDR           0x000040 0x0000000000000040 0x0000000000000040 0x0002d8 0x0002d8 R   0x8
        //   INTERP         0x000318 0x0000000000000318 0x0000000000000318 0x00001c 0x00001c R   0x1
        //   LOAD           0x000000 0x0000000000000000 0x0000000000000000 0x000648 0x000648 R   0x1000
        //   LOAD           0x001000 0x0000000000001000 0x0000000000001000 0x000199 0x000199 R E 0x1000
        //   LOAD           0x002000 0x0000000000002000 0x0000000000002000 0x0000e4 0x0000e4 R   0x1000
        //   LOAD           0x002db0 0x0000000000003db0 0x0000000000003db0 0x000268 0x000270 RW  0x1000
        //   DYNAMIC        0x002dc0 0x0000000000003dc0 0x0000000000003dc0 0x000200 0x000200 RW  0x8
        //   NOTE           0x000338 0x0000000000000338 0x0000000000000338 0x000020 0x000020 R   0x8
        //   NOTE           0x000358 0x0000000000000358 0x0000000000000358 0x000044 0x000044 R   0x4
        //   GNU_PROPERTY   0x000338 0x0000000000000338 0x0000000000000338 0x000020 0x000020 R   0x8
        //   GNU_EH_FRAME   0x002008 0x0000000000002008 0x0000000000002008 0x00002c 0x00002c R   0x4
        //   GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0x10
        //   GNU_RELRO      0x002db0 0x0000000000003db0 0x0000000000003db0 0x000250 0x000250 R   0x1
        stream.seek(SeekFrom::Start(0x2000)).unwrap();
        stream
    }

    fn core_elf_path() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/cli/memfault_core_handler/fixtures/elf-core-runtime-ld-paths.elf")
    }
}
