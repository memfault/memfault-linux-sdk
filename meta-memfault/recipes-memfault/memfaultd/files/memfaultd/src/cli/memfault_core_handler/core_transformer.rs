//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    io::{Read, Seek, SeekFrom},
    sync::mpsc::Receiver,
};
use std::{mem::size_of, time::Duration};

use crate::cli::memfault_core_handler::core_elf_note::{iterate_elf_notes, ElfNote};
use crate::cli::memfault_core_handler::core_reader::CoreReader;
use crate::cli::memfault_core_handler::core_writer::{CoreWriter, SegmentData};
use crate::cli::memfault_core_handler::elf;
use crate::cli::memfault_core_handler::find_dynamic::find_dynamic_linker_ranges;
use crate::cli::memfault_core_handler::find_elf_headers::find_elf_headers_and_build_id_note_ranges;
use crate::cli::memfault_core_handler::find_stack::find_stack;
use crate::cli::memfault_core_handler::memory_range::{merge_memory_ranges, MemoryRange};
use crate::cli::memfault_core_handler::procfs::ProcMaps;
use crate::cli::memfault_core_handler::ElfPtrSize;
use crate::config::CoredumpCaptureStrategy;
use crate::{
    cli::memfault_core_handler::core_elf_memfault_note::{
        write_memfault_metadata_note, CoredumpMetadata,
    },
    mar::LinuxLogsFormat,
};

use elf::program_header::{ProgramHeader, PT_LOAD, PT_NOTE};
use eyre::{eyre, Result};
use libc::{AT_PHDR, AT_PHNUM};
use log::{debug, warn};
use procfs::process::MemoryMap;

use super::{
    core_elf_memfault_note::{write_memfault_debug_data_note, MemfaultMetadataLogs},
    log_wrapper::CAPTURE_LOG_CHANNEL_SIZE,
};

#[derive(Debug)]
pub struct CoreTransformerOptions {
    pub max_size: usize,
    pub capture_strategy: CoredumpCaptureStrategy,
    pub thread_filter_supported: bool,
}

/// Reads segments from core elf stream and memory stream and builds a core new elf file.
pub struct CoreTransformer<R, W, P, M>
where
    R: CoreReader,
    W: CoreWriter,
    P: Read + Seek,
    M: ProcMaps,
{
    core_reader: R,
    core_writer: W,
    proc_mem_stream: P,
    metadata: CoredumpMetadata,
    options: CoreTransformerOptions,
    proc_maps: M,
    log_fetcher: CoreTransformerLogFetcher,
}

impl<R, W, P, M> CoreTransformer<R, W, P, M>
where
    R: CoreReader,
    W: CoreWriter,
    P: Read + Seek,
    M: ProcMaps,
{
    /// Creates an instance of `CoreTransformer` from an input stream and output stream
    pub fn new(
        core_reader: R,
        core_writer: W,
        proc_mem_stream: P,
        options: CoreTransformerOptions,
        metadata: CoredumpMetadata,
        proc_maps: M,
        log_fetcher: CoreTransformerLogFetcher,
    ) -> Result<Self> {
        Ok(Self {
            core_reader,
            core_writer,
            proc_mem_stream,
            metadata,
            options,
            proc_maps,
            log_fetcher,
        })
    }

    /// Reads segments from core elf stream and memory stream and builds a new elf file
    ///
    /// Reads all PT_LOAD and PT_NOTE program headers and their associated segments from the core.
    /// The memory for each PT_LOAD segment will be fetched from `/proc/<pid>/mem` and the
    /// resulting elf file will be written to `output_stream`.
    pub fn run_transformer(&mut self) -> Result<()> {
        let program_headers = self.core_reader.read_program_headers()?;
        let all_notes = self.read_all_note_segments(&program_headers);

        let segments_to_capture = match self.options.capture_strategy {
            CoredumpCaptureStrategy::KernelSelection => {
                self.kernel_selection_segments(&program_headers)
            }
            CoredumpCaptureStrategy::Threads { max_thread_size } => {
                // Fallback to kernel selection if thread filtering is not supported.
                // Support is dictated by whether we have a definition for the platform
                // user regs. See arch.rs for more details.
                if self.options.thread_filter_supported {
                    let mapped_ranges = self.proc_maps.get_process_maps()?;
                    self.threads_segments(&mapped_ranges, &all_notes, max_thread_size)
                } else {
                    // Update metadata so capture strategy is reflected properly in the memfault note.
                    self.metadata.capture_strategy = CoredumpCaptureStrategy::KernelSelection;
                    self.kernel_selection_segments(&program_headers)
                }
            }
        };

        // Always copy over all note segments, regardless of the capturing strategy:
        for (ph, data) in all_notes {
            self.core_writer.add_segment(*ph, SegmentData::Buffer(data));
        }

        for ph in segments_to_capture {
            self.core_writer.add_segment(ph, SegmentData::ProcessMemory);
        }

        self.add_memfault_metadata_note()?;
        self.add_memfault_debug_data_note()?;
        self.check_output_size()?;
        self.core_writer.write()?;

        Ok(())
    }

    /// Add note segments from the core elf to the output elf, verbatim.
    fn read_all_note_segments<'a>(
        &mut self,
        program_headers: &'a [ProgramHeader],
    ) -> Vec<(&'a ProgramHeader, Vec<u8>)> {
        program_headers
            .iter()
            .filter_map(|ph| match ph.p_type {
                PT_NOTE => match self.core_reader.read_segment_data(ph) {
                    Ok(data) => Some((ph, data)),
                    _ => None,
                },
                _ => None,
            })
            .collect::<Vec<_>>()
    }

    /// All load segments from the original/input core elf as provided by the kernel, verbatim.
    fn kernel_selection_segments(
        &mut self,
        program_headers: &[ProgramHeader],
    ) -> Vec<ProgramHeader> {
        program_headers
            .iter()
            .filter_map(|ph| match ph.p_type {
                PT_LOAD => Some(*ph),
                _ => None,
            })
            .collect::<Vec<_>>()
    }

    /// Synthesizes segments for the stacks of all threads and adds them to the output, as well as
    /// any mandatory segments (build id notes, r_debug, etc.).
    fn threads_segments(
        &mut self,
        memory_maps: &[MemoryMap],
        all_notes: &[(&ProgramHeader, Vec<u8>)],
        max_thread_size: usize,
    ) -> Vec<ProgramHeader> {
        let memory_maps_ranges: Vec<MemoryRange> =
            memory_maps.iter().map(MemoryRange::from).collect();

        let parsed_notes = all_notes
            .iter()
            .flat_map(|(_, data)| iterate_elf_notes(data))
            .collect::<Vec<_>>();

        let mut mem_ranges = Vec::new();
        let mut phdr_vaddr = None;
        let mut phdr_num = None;

        for note in &parsed_notes {
            match note {
                ElfNote::ProcessStatus(s) => {
                    if let Some(stack) = find_stack(&s.pr_reg, &memory_maps_ranges, max_thread_size)
                    {
                        mem_ranges.push(stack);
                    } else {
                        warn!("Failed to collect stack for thread: {}", s.pr_pid);
                    }
                }
                ElfNote::Auxv(auxv) => {
                    phdr_num = auxv.find_value(AT_PHNUM);
                    phdr_vaddr = auxv.find_value(AT_PHDR);
                }
                _ => {}
            }
        }

        mem_ranges.extend(
            memory_maps
                .iter()
                .filter(|mmap| mmap.offset == 0)
                .flat_map(
                    |mmap| match self.elf_metadata_ranges_for_mapped_file(mmap.address.0) {
                        Ok(ranges) => ranges,
                        Err(e) => {
                            debug!(
                                "Failed to collect metadata for {:?} @ {:#x}: {}",
                                mmap.pathname, mmap.address.0, e
                            );
                            vec![]
                        }
                    },
                ),
        );

        match (phdr_vaddr, phdr_num) {
            (Some(phdr_vaddr), Some(phdr_num)) => {
                if let Err(e) = find_dynamic_linker_ranges(
                    &mut self.proc_mem_stream,
                    phdr_vaddr,
                    phdr_num,
                    &memory_maps_ranges,
                    &mut mem_ranges,
                ) {
                    warn!("Failed to collect dynamic linker ranges: {}", e);
                }
            }
            _ => {
                warn!("Missing AT_PHDR or AT_PHNUM auxv entry");
            }
        };

        // Merge overlapping memory ranges and turn them into PT_LOAD program headers. As a
        // side-effect, this will also sort the program headers by vaddr.
        let merged_ranges = merge_memory_ranges(mem_ranges);
        merged_ranges.into_iter().map(ProgramHeader::from).collect()
    }

    fn elf_metadata_ranges_for_mapped_file(&mut self, vaddr_base: u64) -> Result<Vec<MemoryRange>> {
        // Ignore unnecessary cast here as it is needed on 32-bit systems.
        #[allow(clippy::unnecessary_cast)]
        self.proc_mem_stream
            .seek(SeekFrom::Start(vaddr_base as u64))?;
        find_elf_headers_and_build_id_note_ranges(
            vaddr_base as ElfPtrSize,
            &mut self.proc_mem_stream,
        )
    }

    /// Check if the output file size exceeds the max size available
    fn check_output_size(&self) -> Result<()> {
        let output_size = self.core_writer.calc_output_size();
        if output_size > self.options.max_size {
            return Err(eyre!(
                "Core file size {} exceeds max size {}",
                output_size,
                self.options.max_size
            ));
        }

        Ok(())
    }

    fn add_memfault_metadata_note(&mut self) -> Result<()> {
        let app_logs = self
            .log_fetcher
            .get_app_logs()
            .map(|logs| MemfaultMetadataLogs::new(logs, LinuxLogsFormat::default()));

        self.metadata.app_logs = app_logs;

        let note_data = write_memfault_metadata_note(&self.metadata)?;
        self.add_memfault_note(note_data)
    }

    fn add_memfault_debug_data_note(&mut self) -> Result<()> {
        let mut capture_logs = self.log_fetcher.get_capture_logs();
        if capture_logs.is_empty() {
            return Ok(());
        }

        if capture_logs.len() == CAPTURE_LOG_CHANNEL_SIZE {
            capture_logs.push("Log overflow, some logs may have been dropped".to_string());
        }

        let buffer = write_memfault_debug_data_note(capture_logs)?;
        self.add_memfault_note(buffer)
    }

    /// Add memfault note to the core elf
    ///
    /// Contains CBOR encoded information about the system capturing the coredump. See
    /// [`CoredumpMetadata`] for more information.
    fn add_memfault_note(&mut self, desc: Vec<u8>) -> Result<()> {
        let program_header = ProgramHeader {
            p_type: PT_NOTE,
            p_filesz: desc.len().try_into()?,
            ..Default::default()
        };
        self.core_writer
            .add_segment(program_header, SegmentData::Buffer(desc));

        Ok(())
    }
}

/// Convenience struct to hold the log receivers for the core transformer.
///
/// This is used to receive logs from the `CoreHandlerLogWrapper` and the the circular buffer
/// capture logs.
#[derive(Debug)]
pub struct CoreTransformerLogFetcher {
    capture_logs_rx: Receiver<String>,
    app_logs_rx: Receiver<Option<Vec<String>>>,
}

impl CoreTransformerLogFetcher {
    /// A timeout of 500s is used to prevent the coredump handler from blocking indefinitely.
    /// This is to prevent a slow http request from blocking the coredump handler indefinitely.
    const APPLICATION_LOGS_TIMEOUT: Duration = Duration::from_millis(500);

    pub fn new(
        capture_logs_rx: Receiver<String>,
        app_logs_rx: Receiver<Option<Vec<String>>>,
    ) -> Self {
        Self {
            capture_logs_rx,
            app_logs_rx,
        }
    }

    /// Fetches all logs that were captured in the coredump handler during execution.
    pub fn get_capture_logs(&self) -> Vec<String> {
        self.capture_logs_rx.try_iter().collect()
    }

    /// Fetches all application logs that were in the circular buffer at time of crash.
    ///
    /// This method will block for a maximum of 500ms before returning. This is to prevent a
    /// slow http request from blocking the coredump handler indefinitely.
    pub fn get_app_logs(&self) -> Option<Vec<String>> {
        let logs = self
            .app_logs_rx
            .recv_timeout(Self::APPLICATION_LOGS_TIMEOUT);

        match logs {
            Ok(logs) => logs,
            Err(e) => {
                debug!("Failed to fetch application logs within timeout: {}", e);
                None
            }
        }
    }
}

impl From<MemoryRange> for ProgramHeader {
    fn from(range: MemoryRange) -> Self {
        ProgramHeader {
            p_type: PT_LOAD,
            p_vaddr: range.start,
            p_filesz: range.size(),
            p_memsz: range.size(),
            p_align: size_of::<ElfPtrSize>() as ElfPtrSize,
            ..Default::default()
        }
    }
}

#[cfg(test)]
mod test {
    use crate::cli::memfault_core_handler::test_utils::{
        FakeProcMaps, FakeProcMem, MockCoreWriter,
    };
    use crate::test_utils::setup_logger;
    use crate::{
        cli::memfault_core_handler::core_reader::CoreReaderImpl, test_utils::set_snapshot_suffix,
    };
    use insta::assert_debug_snapshot;
    use rstest::rstest;
    use std::fs::File;
    use std::path::PathBuf;
    use std::sync::mpsc::sync_channel;

    use super::*;

    #[rstest]
    #[case("kernel_selection", CoredumpCaptureStrategy::KernelSelection, true)]
    #[case("threads_32k", CoredumpCaptureStrategy::Threads { max_thread_size: 32 * 1024 }, true)]
    #[case(
        "threads_32k_no_filter_support",
        CoredumpCaptureStrategy::Threads { max_thread_size: 32 * 1024 },
        false
    )]
    fn test_transform(
        #[case] test_case_name: &str,
        #[case] capture_strategy: CoredumpCaptureStrategy,
        #[case] thread_filter_supported: bool,
        _setup_logger: (),
    ) {
        let input_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/cli/memfault_core_handler/fixtures/elf-core-runtime-ld-paths.elf");
        let input_stream = File::open(&input_path).unwrap();
        let proc_mem_stream = FakeProcMem::new_from_path(&input_path).unwrap();
        let proc_maps = FakeProcMaps::new_from_path(&input_path).unwrap();
        let opts = CoreTransformerOptions {
            max_size: 1024 * 1024,
            capture_strategy,
            thread_filter_supported,
        };
        let metadata = CoredumpMetadata {
            device_id: "12345678".to_string(),
            hardware_version: "evt".to_string(),
            software_type: "main".to_string(),
            software_version: "1.0.0".to_string(),
            sdk_version: "SDK_VERSION".to_string(),
            captured_time_epoch_s: 1234,
            cmd_line: "binary -a -b -c".to_string(),
            capture_strategy,
            app_logs: None,
        };

        let (_capture_logs_tx, capture_logs_rx) = sync_channel(32);
        let (app_logs_tx, app_logs_rx) = sync_channel(1);
        app_logs_tx
            .send(Some(vec![
                "Error 1".to_string(),
                "Error 2".to_string(),
                "Error 3".to_string(),
            ]))
            .unwrap();

        let core_reader = CoreReaderImpl::new(input_stream).unwrap();
        let mut segments = vec![];
        let mock_core_writer = MockCoreWriter::new(&mut segments);
        let log_rx = CoreTransformerLogFetcher {
            capture_logs_rx,
            app_logs_rx,
        };
        let mut transformer = CoreTransformer::new(
            core_reader,
            mock_core_writer,
            proc_mem_stream,
            opts,
            metadata,
            proc_maps,
            log_rx,
        )
        .unwrap();

        transformer.run_transformer().unwrap();

        // Omit the actual data from the notes:
        let segments = segments
            .iter()
            .map(|(ph, seg)| {
                let seg = match seg {
                    SegmentData::ProcessMemory => SegmentData::ProcessMemory,
                    SegmentData::Buffer(_) => SegmentData::Buffer(vec![]),
                };
                (ph, seg)
            })
            .collect::<Vec<_>>();

        set_snapshot_suffix!("{}", test_case_name);
        assert_debug_snapshot!(segments);
    }
}
