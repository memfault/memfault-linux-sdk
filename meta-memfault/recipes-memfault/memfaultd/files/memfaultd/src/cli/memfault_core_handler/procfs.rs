//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::fs::File;
use std::io::{Read, Seek, SeekFrom};

use eyre::{eyre, Result};
use procfs::process::{MemoryMap, MemoryMaps};

use crate::cli::memfault_core_handler::memory_range::MemoryRange;
use crate::cli::memfault_core_handler::ElfPtrSize;

/// Opens /proc/<pid>/mem for reading.
pub fn proc_mem_stream(pid: i32) -> Result<File> {
    let proc_mem_stream = File::open(format!("/proc/{}/mem", pid))?;
    Ok(proc_mem_stream)
}

/// Reads memory from /proc/<pid>/mem into a buffer.
pub fn read_proc_mem<P: Read + Seek>(
    proc_mem_stream: &mut P,
    vaddr: ElfPtrSize,
    size: ElfPtrSize,
) -> Result<Vec<u8>> {
    // Ignore unnecessary cast here as it is needed on 32-bit systems.
    #[allow(clippy::unnecessary_cast)]
    proc_mem_stream.seek(SeekFrom::Start(vaddr as u64))?;
    let mut buf = vec![0; size as usize];
    proc_mem_stream.read_exact(&mut buf)?;
    Ok(buf)
}

pub fn read_proc_cmdline<P: Read>(cmd_line_stream: &mut P) -> Result<String> {
    let mut cmd_line_buf = Vec::new();
    cmd_line_stream.read_to_end(&mut cmd_line_buf)?;

    Ok(String::from_utf8_lossy(&cmd_line_buf).into_owned())
}

/// Wrapper trait for reading /proc/<pid>/maps.
///
/// Provides a useful abstraction that can be mocked out for testing.
pub trait ProcMaps {
    fn get_process_maps(&mut self) -> Result<Vec<MemoryMap>>;
}

/// Direct implementation of ProcMaps that reads from /proc/<pid>/maps file.
///
/// This is the default implementation used in production. It simply reads directly from
/// the file and returns the parsed memory ranges.
#[derive(Debug)]
pub struct ProcMapsImpl {
    pid: i32,
}

impl ProcMapsImpl {
    pub fn new(pid: i32) -> Self {
        Self { pid }
    }
}

impl ProcMaps for ProcMapsImpl {
    fn get_process_maps(&mut self) -> Result<Vec<MemoryMap>> {
        let maps_file_name = format!("/proc/{}/maps", self.pid);

        Ok(MemoryMaps::from_path(maps_file_name)
            .map_err(|e| eyre!("Failed to read /proc/{}/maps: {}", self.pid, e))?
            .memory_maps)
    }
}

impl From<&MemoryMap> for MemoryRange {
    fn from(m: &MemoryMap) -> Self {
        MemoryRange {
            start: m.address.0 as ElfPtrSize,
            end: m.address.1 as ElfPtrSize,
        }
    }
}
