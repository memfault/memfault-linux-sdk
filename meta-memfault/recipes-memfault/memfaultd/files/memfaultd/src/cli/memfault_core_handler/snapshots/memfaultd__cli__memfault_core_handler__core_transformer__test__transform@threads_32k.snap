---
source: memfaultd/src/cli/memfault_core_handler/core_transformer.rs
expression: segments
---
[
    (
        ProgramHeader {
            p_type: "PT_NOTE",
            p_flags: 0x0,
            p_offset: 0x660,
            p_vaddr: 0x0,
            p_paddr: 0x0,
            p_filesz: 0xdec,
            p_memsz: 0x0,
            p_align: 0,
        },
        Buffer(
            [],
        ),
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x5587ae8bd000,
            p_paddr: 0x0,
            p_filesz: 0x39c,
            p_memsz: 0x39c,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x5587ae8c0dc0,
            p_paddr: 0x0,
            p_filesz: 0x200,
            p_memsz: 0x200,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38cb4000,
            p_paddr: 0x0,
            p_filesz: 0x3b4,
            p_memsz: 0x3b4,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38e99000,
            p_paddr: 0x0,
            p_filesz: 0x30c,
            p_memsz: 0x30c,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38f781d0,
            p_paddr: 0x0,
            p_filesz: 0xc,
            p_memsz: 0xc,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38f781e0,
            p_paddr: 0x0,
            p_filesz: 0x28,
            p_memsz: 0x28,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38f78740,
            p_paddr: 0x0,
            p_filesz: 0x20,
            p_memsz: 0x20,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38f78760,
            p_paddr: 0x0,
            p_filesz: 0x28,
            p_memsz: 0x28,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38f7a000,
            p_paddr: 0x0,
            p_filesz: 0x25c,
            p_memsz: 0x25c,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38facad0,
            p_paddr: 0x0,
            p_filesz: 0x28,
            p_memsz: 0x28,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38fad118,
            p_paddr: 0x0,
            p_filesz: 0x28,
            p_memsz: 0x28,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38fad2e0,
            p_paddr: 0x0,
            p_filesz: 0x28,
            p_memsz: 0x28,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38fad8a0,
            p_paddr: 0x0,
            p_filesz: 0x1,
            p_memsz: 0x1,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38fad8b0,
            p_paddr: 0x0,
            p_filesz: 0x28,
            p_memsz: 0x28,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7ffe4ba7a3c0,
            p_paddr: 0x0,
            p_filesz: 0x1c40,
            p_memsz: 0x1c40,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7ffe4bb1f000,
            p_paddr: 0x0,
            p_filesz: 0x554,
            p_memsz: 0x554,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_NOTE",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x0,
            p_paddr: 0x0,
            p_filesz: 0xc8,
            p_memsz: 0x0,
            p_align: 0,
        },
        Buffer(
            [],
        ),
    ),
]
